### ISSUE TITLE
mouse over tooltip for overlay objects

### ISSUE THREAD

**Issue by gijzelaerr** [ericmandel/js9#7]
Hi!

maybe this is already possible and I have not found this, but it would be very useful if it was possible to be able to add have mouse over/hover tooltip for overlay objects. That we a user can see more info (like source ID) if you move the mouse over the overlay. Maybe even clickable to move to a an other page related to the source.

**Comment by ericmandel** [comment]
@gijzelaerr I'd be happy to work with you on adding this functionality. We have events turned off in JS9.Catalogs.opts by default ... mainly because we didn't quite know what was wanted. You can turn it on by setting these options before creating the layer:
````
JS9.Catalogs.opts.selectable = true
JS9.Catalogs.opts.evented = true
````
Then if you load add objects to a layer, you can double-click on them and the region dialog box come up with region info ... but not the info you want. Perhaps we should think about extending that. You also can move the objects around (which is not good for catalogs in general but ...)

What sort of functionality should we add? It sounds like you want to be able to define a callback for onhover and get passed the information for that catalog object, at least to start.

**Comment by gijzelaerr** [comment]
Hi Eric!

I'm integrating JS9 in the webinteface (banana) of our transient detection pipeline (TraP).

https://github.com/transientskp/banana

In banana we visualize a big list of found point sources in an overlay of the image using DS9. Each of the source has a unique ID and also a unique page containing more information about the source.  Without a tooltip or to make the circle clickable it gets quite hard to know which source is which.

![js9](https://cloud.githubusercontent.com/assets/326308/10169346/e9b50452-66d0-11e5-85b0-5183b5133e73.png)

**Comment by ericmandel** [comment]
That's a really nice app, so let's try to make it better. What I think you need is:

1. ability to specify a callback when you hover over a shape, the callback returning JS9 shape data and whatever catalog data you passed to JS9 so that you can bring up a tooltip with whatever you want in it.

2. ability to specify a callback for ... double-click looks good for your purpose, so that you can retrieve and display more information in a light window, or perhaps highlight a row in your table.

In essence, we need to allow callbacks to be attached to the various events associated with each shape. Probably they get defined as optional properties in the opts object when you create the layer.

I can look into this after we release v1.6, which is scheduled for Oct 13. My one hesitation is that I can't recall why it's not implemented already, since it was talked about internally! But let me have a look.

**Comment by gijzelaerr** [comment]
Eric thanks for your answer and willingness to improve this! JS9 is really one of the tools I was looking for for a long time.

Maybe also the highcharts API can be of inspiration for us.

http://api.highcharts.com/highcharts

Basically you can supply a small peace of javascript/html what will render inside a div rendered next to the mouse when you hover over a object of interest.

**Comment by ericmandel** [comment]
Hi Gijs,

highcharts looks really nice (the plotting supports error bars, which is unusual), but unfortunately, JS9 is being developed by NASA/Smithsonian Institution so highcharts will not be free for us. But  you might be able to use it for your particular app inside a JS9-supplied callback.

**Comment by gijzelaerr** [comment]
Hi Eric,

I was not suggesting to use highcharts, just get inspired by how they
enable the developer to create a custom mouseover tooltip.

2015-09-30 15:00 GMT+02:00 Eric Mandel <notifications@github.com>:

> Hi Gijs,
>
> highcharts looks really nice (the plotting supports error bars, which is
> unusual), but unfortunately, JS9 is being developed by NASA/Smithsonian
> Institution so highcharts will not be free for us. But you might be able to
> use it for your particular app inside a JS9-supplied callback.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ericmandel/js9/issues/7#issuecomment-144388722>.
>



-- 
Gijs Molenaar
http://pythonic.nl

**Comment by ericmandel** [comment]
Oh, now I understand, thanks. But I do wish we had the chance to use that module, it looks very nice.

**Comment by ericmandel** [comment]
@gijzelaerr I have implemented a first offering of tooltips and event handlers for shape layers. The code has been uploaded into the JS9 repository. You can view a demo [here](http://js9.si.edu/js9/js9cat.html).

There are two levels of support:

At a primitive level, you can specify onmouse[up,down,move,over,out] callbacks via properties in the JS9.NewShapeLayer() opts object, allowing you to do your own processing when a event happens on a shape. The calling sequence for these callbacks is:
````
    mousex(im, xreg, evt)
````
where im is the current image handle, xreg is the shape object, and evt is the original event. The contents of the xreg object is described in the documentation for JS9.GetShapes().

At a higher convenience level, you can specify a tooltip "format string" which can reference data associated with the current image (im) or the current shape (xreg). When you hover over an object, a tooltip string is generated  on the fly from the format string  (macro-expanding the im and xreg references) and displayed. In this way, you can choose what parts of the data to display in the tooltip or during even processing. See the demo for a detailed explanation.

One important thing to note for your specific application is that you can pass data to associate with each object when you call JS9.AddShapes(). This data will be accessible in xreg in xreg.data, and you can use it in the event processing calls, or even in the tooltip. Just read the demo code to see how it's done.

Let me know how you see things ...

**Comment by gijzelaerr** [comment]
Thanks Erik! Great work. I will have a look.

**Comment by ericmandel** [comment]
@gijzelaerr   Hi Gijs, and while you are doing that, I am looking into the possibility of defining a callback for when you select a group of objects, which, when fired, would get passed an array of xreg objects to process. Akin to what glue does in Python, but much simpler.

Also, I just fixed js9.css, so if you already cloned the new stuff, you should update. The tooltip was not properly hidden on init ... oops.

**Comment by ericmandel** [comment]
@gijzelaerr Hi Gijs, I implemented support for an ongroupcreate event callback, which fires when you use the mouse to create a temporary group of shapes (using the Mac cmd key or Linux ctl key). The callback is like the single-shape callbacks above, but returns an array of xreg shapes instead of a single one. This ought to open up a lot of doors for you to process objects ...

New code has been committed to the repository and I've updated the JS9 Web site. See [http://js9.si.edu/js9/js9cat.html](http://js9.si.edu/js9/js9cat.html]) for the demo and documentation (formally documented in the publicapi html page under JS9.NewShapeLayer).

I really like the way the implementation came out easily from the overall structure, but let me know what you find ...
